Decompress FITS files in memory

saimn opened the pull request:
Fix #14855

<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is to address ...

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #<Issue Number>

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

astrofrog commented on astropy/io/fits/hdu/hdulist.py:
```diff
@@ -184,6 +185,14 @@ def fitsopen(
         .. versionadded:: 5.2
 
+    decompress_in_memory : bool, optional
+        By default blocks are decompressed on the fly, when reading them.  This
```
I think the use of 'blocks' isn't necessarily going to be clear to the average user. Can we instead say something like 'By default, files are decompressed progressively depending on what data is needed' (not attached to this wording, but just not using blocks - which to me makes me think of tiles in tiled image compression)

astrofrog commented on docs/changes/io.fits/15501.feature.rst:
```diff
@@ -0,0 +1,4 @@
+Add new option ``decompress_in_memory=False`` to ``fits.open``, to decompress
+the whole file in memory at once, instead of decompressing small blocks on the
```
maybe you could say 'instead of decompressing the file progressively as data is needed' or similar, again avoiding the term block.

astrofrog reviewed (approved):
Looks good to me and this will be very useful! Just a couple of small comments below on terminology but approving since not critical if you don't have time.

astrofrog commented:
Actually could you add a 'What's New' entry for this?

saimn commented:
@astrofrog - I was thinking "FITS blocks" but I agree it's not so clear :)
Updated with your suggestion (thanks!), and also added a small whatsnew section.
